File: trunk/TeoriaDeAlgoritmos/src/main/java/ar/fi/uba/tda/colecciones/ListaEnlazada.java
package ar.fi.uba.tda.colecciones;

import java.util.Iterator;

public class ListaEnlazada<T> {
	
	private Elemento header = new Elemento(null, null, null);
	private Integer tamanio = 0;

	public ListaEnlazada () {
		
		header.siguiente = header.anterior = header;
	}
	
	public Boolean vacia() {
		return tamanio == 0;
	}

	public T primero() {
		return header.siguiente.elemento;
	}

	public T ultimo() {
		return header.anterior.elemento;
	}
	
	public void agregar(T elemento) {
		agregarAntes(elemento, header);
	}
	
	private void agregarAntes(T elemento, Elemento siguiente) {
		
		Elemento nuevo = new Elemento(elemento, siguiente, siguiente.anterior);
		nuevo.anterior.siguiente = nuevo;
		nuevo.siguiente.anterior = nuevo;
		tamanio++;
	}
	
	public Integer tamanio() {
		return tamanio;
	}

	public Iterator<T> iterador() {
		return new IteradorListaEnlazada();
	}
	
	private class IteradorListaEnlazada implements Iterator<T> {

		private int siguiente;
		private Elemento siguienteElemento;
		
		public IteradorListaEnlazada () {
		
			siguiente = 0;
			siguienteElemento = header;
		}
		
		@Override
		public boolean hasNext() {
			return siguiente != tamanio;
		}

		@Override
		public T next() {
			
			T aRetornar = null;
			
			if (siguiente < tamanio) {
				
				aRetornar = siguienteElemento.siguiente.elemento;
				siguienteElemento = siguienteElemento.siguiente;
				siguiente++;
			}
			
			return aRetornar;
		}

		@Override
		public void remove() {
			
		}
		
	}
	
	private class Elemento {
		
		T elemento;
		Elemento anterior;
		Elemento siguiente;
		
		public Elemento(T elemento, Elemento siguiente, Elemento anterior) {
			
			this.elemento = elemento;
			this.anterior = anterior;
			this.siguiente = siguiente;
		}

	}

	public boolean contiene(T datoBuscado) {
		
		return obtener(datoBuscado) != null;
	}

	public T obtener(T buscado) {
		
		T encontrado = null;
		Iterator<T> iterador = this.iterador();
		
		while (encontrado == null && iterador.hasNext()) {
			
			T item = iterador.next();
			
			if (item.equals(buscado)) {
				encontrado = item;
			}
		}
		
		return encontrado;
	}

}

File: Robustez/Robustez/ListaEnlazada.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Robustez
{
    class ListaEnlazada<T>
    {
        private static Elemento<T> header = new Elemento<T>(default(T), null, null);
        private static Int32 tamanio = 0;

        public static Int32 Tamanio
        {
            get { return ListaEnlazada < T >.tamanio; }
            set { ListaEnlazada<T>.tamanio = value; }
        }

	    public ListaEnlazada () {

		    header.Siguiente = header.Anterior = header;
	    }
	
	    public bool vacia() {
		    return true;
	    }

	    public T primero() {
            return header.Siguiente.Vertice;
	    }

	    public T ultimo() {
		    return header.Anterior.Vertice;
	    }
	
	    public void agregar(T elemento) {
		    agregarAntes(elemento, header);
	    }
        public Int32 GetTamanio()
        {
            return tamanio;
        }
	
	    private void agregarAntes(T elemento, Elemento<T> siguiente) {
		
		    Elemento<T> nuevo = new Elemento<T>(elemento, siguiente, siguiente.Anterior);
		    nuevo.Anterior.Siguiente = nuevo;
            nuevo.Siguiente.Anterior = nuevo;
		    tamanio++;
	    }

        public IteradorListaEnlazada iterador()
        {
		    return new IteradorListaEnlazada();
	    }
	
	    public class IteradorListaEnlazada  {

		    private int siguiente;
		    private Elemento<T> siguienteElemento;
		
		    public IteradorListaEnlazada () {
		
			    siguiente = 0;
			    siguienteElemento = header;
		    }
		

		    public bool hasNext() {
			    return siguiente != tamanio;
		    }

		    public T next() {
			
			    T aRetornar = default(T);
			
			    if (siguiente < Tamanio) {
				
				    aRetornar = siguienteElemento.Siguiente.Vertice;
				    siguienteElemento = siguienteElemento.Siguiente;
				    siguiente++;
			    }
			
			    return aRetornar;
		    }


		    public void remove() {
			
		    }

            /*
            #region IEnumerable<T> Members

            public IEnumerator<T> GetEnumerator()
            {
                throw new Exception("The method or operation is not implemented.");
            }

            #endregion

            #region IEnumerable Members

            System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
            {
                throw new Exception("The method or operation is not implemented.");
            }

            #endregion
             */
        }
	
	    private class Elemento<T> {

            private T vertice;
            private Elemento<T> anterior;
            private Elemento<T> siguiente;

            public T Vertice
            {
                get { return vertice; }
                set { vertice = value; }
            }
            
            public Elemento<T> Anterior
            {
                get { return anterior; }
                set { anterior = value; }
            }
            

            public Elemento<T> Siguiente
            {
                get { return siguiente; }
                set { siguiente = value; }
            }
		
		    public Elemento(T elemento, Elemento<T> siguiente, Elemento<T> anterior) {
			
			    this.vertice = elemento;
			    this.anterior = anterior;
			    this.siguiente = siguiente;
		    }

	    }

	    public bool contiene(T datoBuscado) {
		
		    bool encontrado = false;
            IteradorListaEnlazada iterador = this.iterador();
		
		    while (!encontrado && iterador.hasNext()) {
			
			    T item = iterador.next();
			
			    if (item.Equals(datoBuscado)) {
				    encontrado = true;
			    }
		    }
		
		    return encontrado;
	    }

       
    }
}
